src/numeric.rb

```diff
@@ -158,71 +162,17 @@ def step(to_pos = nil, by_pos = nil, by: nil, to: nil)
       raise ArgumentError, 'by is given twice'
     end
 
-    by ||= by_pos || 1
+    by ||= by_pos
     to ||= to_pos
 
-    if !by.is_a?(Numeric) && by.respond_to?(:to_int)
-      by = by.to_int
-    end
-
-    # This is here for compatiblity with ruby.
-    # When `by` is a non-numeric that cannot be compared with zero, this
-    # should raise an ArgumentError.
-    ascending = by > 0
-
-    if !by.is_a?(Numeric)
-      raise TypeError, "no implicit conversion of #{by.class} into Integer"
-    end
-
     if by == 0
-      raise ArgumentError, 'step can\'t be 0'
-    end
-
-    # Special handling for infinite values
-    if by.is_a?(Float) && by.infinite?
-      if ascending ? self <= to : self >= to
-        yield to_f
-      end
-      return self
+      raise ArgumentError, "step can't be 0"
     end
 
-    if to.is_a?(Float) && to.infinite?
-      return self
-    end
-
-    # If any argument is a float, we want to use floats from here on
-    if is_a?(Float) || by.is_a?(Float) || to.is_a?(Float)
-      by = by.to_f
-      to = to.to_f
-    end
-
-    # Calculate how many times we have to iterate a step
-    step_count =
-      if by.is_a?(Integer)
-        n = to - self
-        ascending ? n : -n
-      else
-        n = (to - self) / by
-        (n + n * Float::EPSILON).floor
-      end
-
-    if step_count.negative?
-      return self
-    end
+    enumerator = __call__('Enumerator::ArithmeticSequenceObject::from_numeric', self, to, by)
```

Comment:
You might want to look at the new `__bind__` macro, which serves a similar purpose, but would allow you to call it like a normal Ruby method. The `__call__` macro was originally designed for calling C APIs.

Reply:
You mean __bind_method__? This defines the method on the object though which is not want I want here, right? I just want to call the method and use the return value.

Reply:
Sorry, yes, `__bind_method__`. Well you can use `__bind_method__` instead of `__function__`, since it makes a binding to the Ruby object. Then you don't have to use `__call__` here.

Reply:
:thinking: Though... the ergonomics of getting the arguments out are little bit worse with `__bind_method__`, so I'm not recommending you change anything here. I just wanted you to be aware of it. :hugs: